KEP-2033: promote KubeletInUserNamespace feature to beta (v1.37)

AkihiroSuda opened the pull request:
Depends on:
- [x] https://github.com/kubernetes/enhancements/pull/5388
- [x] https://github.com/kubernetes/enhancements/pull/5694
- [x] https://github.com/kubernetes/enhancements/pull/6114
- [x] https://github.com/kubernetes/enhancements/pull/6256

- - -
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
/kind api-change

#### What this PR does / why we need it:
Promote the `KubeletInUserNamespace` feature to beta, with addition of `NodeSystemInfo.RunningInUserNamespace *bool`, as discussed in the `kubernetes/enhancements` PRs above.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
KEP: https://github.com/kubernetes/enhancements/issues/2033

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
KEP-2033: promote KubeletInUserNamespace feature to beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2033
```

AkihiroSuda committed:
api: KEP-2033: add NodeSystemInfo.RunningInUserNamespace

This commit modifies the following files:

- pkg/apis/core/types.go
- staging/src/k8s.io/api/core/v1/types.go

Other changes were auto-generated by running `make update`.

This commit only adds the field definition.
The implementation will be provided in a follow-up commit.

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
kubelet: KEP-2033: implement NodeSystemInfo.RunningInUserNamespace

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
KEP-2033: promote KubeletInUserNamespace feature to beta

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
/label api-review

HirazawaUi reviewed (commented):
Do we need to add a test case for it to verify that the `NodeSystemInfo.RunningInUserNS` field will be correctly set?

Although time is tight, it seems we could add a pre-submit test job similar to ci-[kubernetes-e2e-kind-rootless](https://github.com/kubernetes/test-infra/blob/0468667c27529d1b65fce5be8c9a3b71b733e042/config/jobs/kubernetes/sig-testing/kubernetes-kind-ci.yaml#L456-L458), which would only run tests related to the `KubeletInUserNamespace` feature gate. I believe it would work well :)

Although I checked the original KEP and found no mention of this requirement, in the latest update to the KEP, we added a field to the node API. Therefore, I believe adding an e2e test is justified.

HirazawaUi commented:
> Not sure.
> [The current test](https://github.com/kubernetes/test-infra/blob/0468667c27529d1b65fce5be8c9a3b71b733e042/config/jobs/kubernetes/sig-testing/kubernetes-kind-ci.yaml#L456-L458) uses `--skip-regex='\[Environment:NotInUserNS\]` to skip the tests that do not make sense in UserNS.
> However, we can't just introduce the opposite one like `--skip-regex='\[Environment:InUserNS\]`, as it would incur modifying tons of test scripts.

I think we can use `--focus-regex=\[KubeletInUserNamespace\]` to run only the tests related to the `KubeletInUserNamespace` feature gate (if we've added it), without needing to worry about environment variables like InUserNS or NotInUserNS.

AkihiroSuda committed:
test/e2e/node: split `pidOfKubelet` helper function

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda committed:
e2e: KEP-2033: test KubeletInUserNamespace

To verify:

```
dockerd-rootless-setuptool.sh install
kubetest2-kind --build --up --down --test=ginkgo -- --use-built-binaries  --focus-regex='\[Feature:KubeletInUserNamespace\]'
```

Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>

AkihiroSuda commented:
`[It] [sig-apps] Deployment iterative rollouts should eventually progress`  failures do not seem relevant 
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-e2e-kind/1986241560492118016
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-e2e-kind-alpha-beta-enabled/1986241561452613632
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-e2e-kind-alpha-beta-features/1986241560664084480

SergeyKanzhelev commented on pkg/apis/core/types.go:
```diff
@@ -5538,6 +5538,10 @@ type NodeSystemInfo struct {
+	// Whether the node is running in a user namespace.
+	// +featureGate=KubeletInUserNamespace
+	// +optional
+	RunningInUserNS *bool
```
sorry, I didn't catch this in KEP review. Typically a new API field will mean staying in alpha for another release. I think it may be the most reasonable thing here given where we are in release cycle. Do you think you can update the target stage for this KEP?

AkihiroSuda replied:
> Typically a new API field will mean staying in alpha for another release.

I hope that this general rule doesn't necessarily apply for adding a single boolean to a status object.

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -6618,6 +6618,10 @@ type NodeSystemInfo struct {
 	// Swap Info reported by the node.
 	Swap *NodeSwapStatus `json:"swap,omitempty" protobuf:"bytes,11,opt,name=swap"`
+	// Whether the node is running in a user namespace.
+	// +featureGate=KubeletInUserNamespace
```
clear this field in `nodeStatusStrategy#PrepareForUpdate` if the feature is disabled?

```go
	if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletInUserNamespace) && oldNode.Status.NodeInfo.RunningInUserNamespace == nil {
		newNode.Status.NodeInfo.RunningInUserNamespace = nil
	}
```

AkihiroSuda commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -6618,6 +6618,10 @@ type NodeSystemInfo struct {
+	// Whether the node is running in a user namespace.
+	// +featureGate=KubeletInUserNamespace
+	// +optional
+	RunningInUserNS *bool `json:"runningInUserNS,omitempty" protobuf:"varint,12,opt,name=runningInUserNS"`
```
> so runningInUserNamespace would be more similar.

I abbreviated it `runningInUserNS` because I felt `runnningInUserNamespace` is too long, but not so long as `requiredDuringSchedulingIgnoredDuringExecution` etc., so I can rename it to `runningInUserNamespace`

> I don't have a lot of context for what is intended to go into the NodeSystemInfo struct... the kubelet doesn't report other details of how it is running (like it's own uid, etc). What are things watching the Node API object going to do with this status field?

This was suggested in https://github.com/kubernetes/enhancements/pull/5388#discussion_r2414748084

Likely to be watched by humans for diagnostic purposes.

> It looks like the KubeletInUserNamespace feature mostly makes the kubelet ignore errors trying to set sysctls and set up the oomwatcher?

Yes.

dchen1107 commented:
/approve

SergeyKanzhelev commented:
/skip
/milestone v1.35

(as per https://kubernetes.slack.com/archives/C2C40FMNF/p1762645171061649)

/lgtm

/hold

(unhold at will. Holding to prevent accidental merges)

API review is still required. @liggitt or @thockin?

seans3 commented:
/remove-sig api-machinery

AkihiroSuda commented on staging/src/k8s.io/api/core/v1/well_known_labels.go:
```diff
@@ -71,4 +71,8 @@ const (
+
+	// LabelRunningInUserNamespace is added by kubelet to a node to indicate whether the kubelet is running in a user namespace.
+	// The value is either "true" or "false".
+	LabelRunningInUserNamespace = "kubernetes.io/running-in-user-namespace"
```
Changed to use the `node.kubernetes.io/` prefix

aaron-prindle commented:
/triage accepted

AkihiroSuda commented:
@liggitt @SergeyKanzhelev @thockin Could you take a look?

AkihiroSuda commented:
Rebased

AkihiroSuda commented:
/retest

AkihiroSuda commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -245,6 +245,10 @@ func (kl *Kubelet) updateDefaultLabels(initialNode, existingNode *v1.Node) bool
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletInUserNamespace) {
+		defaultLabels = append(defaultLabels, v1.LabelRunningInUserNamespace)
```
This was documented in:
- https://github.com/kubernetes/enhancements/pull/6114

SergeyKanzhelev commented on pkg/kubelet/kubelet_node_status_linux.go:
```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2025 The Kubernetes Authors.
```
remove the year...

AkihiroSuda replied:
done

SergeyKanzhelev reviewed (approved):
/lgtm

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/assign deads2k

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

AkihiroSuda commented:
The failure in `pull-kubernetes-unit-windows-master` seems irrelevant


> [sig-scheduling] k8s.io/kubernetes/pkg/scheduler: framework
>> {Failed  === RUN   TestPodGroupMemberPodsOrderingFunc/older_timestamp_comes_first     types_test.go:190: Unexpected result, want -1, got 0 --- FAIL: TestPodGroupMemberPodsOrderingFunc/older_timestamp_comes_first (0.00s)  === RUN   TestPodGroupMemberPodsOrderingFunc/newer_timestamp_comes_second     types_test.go:190: Unexpected result, want 1, got 0 --- FAIL: TestPodGroupMemberPodsOrderingFunc/newer_timestamp_comes_second (0.00s)  === RUN   TestPodGroupMemberPodsOrderingFunc --- FAIL: TestPodGroupMemberPodsOrderingFunc (0.00s) }


https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-unit-windows-master/2061704821106282496

SergeyKanzhelev commented:
/skip

AkihiroSuda commented:
@SergeyKanzhelev Could you unhold?
https://github.com/kubernetes/kubernetes/pull/134639#issuecomment-3509085668

SergeyKanzhelev commented:
/unhold

HirazawaUi commented:
> /unhold

~/hold cancel~

This PR is not in _hold_ state.

dims commented:
/assign @thockin @liggitt

AkihiroSuda commented:
> /assign @thockin @liggitt

@thockin @liggitt Could you take a look? Thanks

kishen-v commented:
Hey @thockin @liggitt, 
I'm currently wrangling this KEP and would appreciate if you can help with the review when time permits.

Thank you!

AkihiroSuda commented:
Rebased

@SergeyKanzhelev Could you lgtm again?

SergeyKanzhelev commented:
> @SergeyKanzhelev Could you lgtm again?

fix linter and ping again. Thx

AkihiroSuda commented:
/retest

AkihiroSuda commented:
> > @SergeyKanzhelev Could you lgtm again?
> 
> fix linter and ping again. Thx

Done, thanks @SergeyKanzhelev

SergeyKanzhelev commented:
/lgtm

SergeyKanzhelev commented:
@liggitt the node label seems to be the last unresolved comment. The node label will be used on kubelets running in user namespace. As we will see more of those, I believe some system daemonsets will start excluding those by default and will beneit to have a built-in label.

liggitt commented:
> The node label will be used on kubelets running in user namespace. As we will see more of those, I believe some system daemonsets will start excluding those by default and will beneit to have a built-in label.

The KEP indicates this as the rationale for the label:

> Nodes will have node.kubernetes.io/running-in-user-namespace: <BOOL> labels. An operator may utilize this label to avoid scheduling workloads that require real root privileges (e.g., CNI plugin installer) on nodes running in user namespaces.

Are those types of pods already able to be determined based on the shape of the pod (e.g. `hostUsers: null` or `hostUsers: true`)? Doesn't the scheduler already have visibility in the node status to whether the kubelet is able to run pods like that? I'm a little confused why the existing node status field is insufficient.

liggitt commented:
https://github.com/kubernetes/enhancements/pull/5694/changes changed the proposed API surface from just a node status field to include a selectable label. @thockin acked it but I don't really understand the use case for the label surface area.

AkihiroSuda commented:
> Are those types of pods already able to be determined based on the shape of the pod (e.g. hostUsers: null or hostUsers: true)? 

`hostUsers` (pods in UserNS, gated by [`NodeRuntimeHandlerFeatures.userNamespaces`](https://kubernetes.io/docs/reference/kubernetes-api/core/node-v1/#NodeRuntimeHandlerFeatures)) is irrelevant to this KEP (entire node in UserNS).

> a selectable label.

For example, the label can be appended to Calico's daemonset which requires the nodes to be running as the real root
https://github.com/projectcalico/calico/blob/v3.32.1/charts/calico/templates/calico-node.yaml#L24-L29
```yaml
      nodeSelector:
        kubernetes.io/os: linux
{{- if .Values.flannel_migration}}
        # Only run Calico on nodes that have been migrated.
        projectcalico.org/node-network-during-migration: calico
{{- end}}
        node.kubernetes.io/running-in-user-namespace: false # NEW
```

liggitt commented:
> For example, the label can be appended to Calico's daemonset which requires the nodes to be running as the real root

so will calico just ... break on kubelets running this way?

AkihiroSuda commented:
> > For example, the label can be appended to Calico's daemonset which requires the nodes to be running as the real root
> 
> so will calico just ... break on kubelets running this way?

Yes, w.r.t. its default Bird BGP mode, as kubelets running this way are running in an NetNS associated with the UserNS.

I heard VXLAN mode still works:
- https://github.com/rootless-containers/usernetes/pull/386

thockin commented on pkg/apis/core/types.go:
```diff
@@ -5612,6 +5612,10 @@ type NodeSystemInfo struct {
+	// Whether the node is running in a user namespace.
+	// +featureGate=KubeletInUserNamespace
+	// +optional
+	RunningInUserNamespace *bool
```
Bool fields often end up being something we regret.  Before I approve this, did we consider options?

Also, the name is long, but I won't demand a change there.

E.g.

```
type UserNSMode string

const (
    UserNSRoot UserNSMode = "Root"
    UserNSIsolated UserNSMOde = "Isolated"
)

...

UserNSMode *UserNSMode
```

I'm not actually sure it's better -- it seems unlikely that we end up with a 3rd option (but we have said that in the past about other fields).  Would there possibly be an option in the future to run this in a nested userns where we actually care about the depth of the nesting?

```
// 0 means the root userNS, any other value means that kubelet is running in a userNS
UserNSDepth *int32 
```

?

thockin replied:
Also - should this have a default value?  Or is it intentional to leave `nil` == "don't know".  Comment.

AkihiroSuda replied:
> Bool fields often end up being something we regret.

When a need arises to support non-boolean values, we can introduce `BoolOrString` type as in https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString

> UserNSDepth

Unimplementable, as a process running in the non-initial UserNS can't inspect its own current depth.

> Or is it intentional to leave nil == "don't know".

Intentional.

SergeyKanzhelev replied:
If we will go with string - it will be great to have it defaulted so we do not need to say "Root" for all nodes as it is a default that will be used in majority of cases. The non-Root is an exception

thockin replied:
> When a need arises to support non-boolean values, we can introduce BoolOrString type

nonononono, that is generally seen as a mistake now :)

> Unimplementable, as a process running in the non-initial UserNS can't inspect its own current depth.

If it were important we would find a way to get that information in.  I am not convinced it is important, though

I don't see a likely path to a third value, so I am ok with this.

thockin commented on staging/src/k8s.io/api/core/v1/well_known_labels.go:
```diff
@@ -71,4 +71,8 @@ const (
 	// LabelMetadataName is the label name which, in-tree, is used to automatically label namespaces, so they can be selected easily by tools which require definitive labels
 	LabelMetadataName = "kubernetes.io/metadata.name"
+
+	// LabelRunningInUserNamespace is added by kubelet to a node to indicate whether the kubelet is running in a user namespace.
```
Why do we need this AND a status?

AkihiroSuda replied:
https://pkg.go.dev/k8s.io/api/core/v1#NodeSystemInfo seems designed to provide the comprehensive view of the system status.
It already has redundant properties (OperatingSystem, Architecture).

thockin replied:
...because those often need to be used in selectors -- do you think node-selectors expressing "I need nodes with userns" is going to be common for this feature?

thockin replied:
I see the other discussion in the main conversataion, will respond there

AkihiroSuda commented:
Rebased again.

I have initially submitted the beta KEP (https://github.com/kubernetes/enhancements/pull/5388) in June 2025 and already missed several milestones.
I hope I can finally get this accepted in v1.37.

AkihiroSuda commented:
/retest

AkihiroSuda commented:
/retest

AkihiroSuda commented:
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> pull-kubernetes-e2enode-windows | 2c7787371b7a65fe23720cdb6026ffb68bc63c9d | [link](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-e2enode-windows/2079623010360233984) | false | `/test pull-kubernetes-e2enode-windows`
> pull-kubernetes-unit-windows-master | 2c7787371b7a65fe23720cdb6026ffb68bc63c9d | [link](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134639/pull-kubernetes-unit-windows-master/2079623010473480192) | false | `/test pull-kubernetes-unit-windows-master`

These failures do not seem related to my PR.

/retest

AkihiroSuda commented:
@liggitt @thockin @SergeyKanzhelev
Is there any remaining concern?

liggitt commented:
> For example, the label can be appended to Calico's daemonset which requires the nodes to be running as the real root https://github.com/projectcalico/calico/blob/v3.32.1/charts/calico/templates/calico-node.yaml#L24-L29
> 
> ```yaml
>       nodeSelector:
>         kubernetes.io/os: linux
> {{- if .Values.flannel_migration}}
>         # Only run Calico on nodes that have been migrated.
>         projectcalico.org/node-network-during-migration: calico
> {{- end}}
>         node.kubernetes.io/running-in-user-namespace: false # NEW
> ```

Adding `node.kubernetes.io/running-in-user-namespace: false` will make that daemonset fail to run on <1.37 kubelets (or any GA-only cluster until this feature graduates). That's a painful rollout story.

It seems like this configuration makes the kubelet newly incapable of running certain types of pods that previously worked, and the selector is being added to put the responsibility on authors of those pods to know they are affected and opt out of running on kubelets in this mode? I'm surprised at that approach, given the shift to make pod / node matchups automatically direct pods to nodes whose versions and configurations can handle them (https://kubernetes.io/docs/concepts/scheduling-eviction/node-declared-features/). I would have expected something introspecting the node configuration (like the status field indicating the kubelet is running in a user namespace) to inform the scheduler it could not schedule host user pods onto that node.

thockin commented:
Jordan got the insight before me -- `node.kubernetes.io/running-in-user-namespace: false` will not be satisfied on anything less than 1.37 and (sadly) Pod NodeSelector is the old map-style, so we can't even discuss how to express it better.

In short, this feature will break a part of our ecosystem which has non-trivial adoption.  And worse, it will do it silently.  Any user who runs Calico in BGP mode today who upgrades their cluster (or worse, whose managed provider does so) will just stop working.

That's a breaking change.

I don't think we can change the default enablement to `true` until we have a more complete story -- am I over-reacting?

AkihiroSuda commented:
I'd suggest Calico etc. to use nodeSeletor only after the EOL of the release that does not have this feature as GA.

Until then they could just use nodeAffinity with a NotIn operator to avoid having a breaking change

thockin commented:
This is still going to hard-break them and their users.  I think it needs to be opt-in?

AkihiroSuda commented:
> This is still going to hard-break them and their users. I think it needs to be opt-in?

The feature gate does not automatically put kubelet into UserNS in the first place.
It just allows the user to opt-in to put it in UserNS.

thockin commented:
Trying to recap slack discussion.

Rolling out this feature is hard.  Cluster admin is expected to know if/which pods will fail with this feature.  Adding a label lets them steer away from nodes with this feature, but for things like Calico that doesn't help.  

I don't see people really wanting to have different kubelet configs on different nodes, and networking doesn't mix-and-match anyway. 

I don't hate the label, I guess, but it seems like a really bad precedent -- every new node feature ends up with node labels? 

This isn't a problem with the feature, but with how we roll it out

thockin commented:
I propose to drop the label - I don't think it actually helps the rollout problem.  Focus for GA needs to be how we help people safely roll it out.  If we decide the label is right, we can bring it back

AkihiroSuda commented:
> I propose to drop the label

Done.
k/e PR:
- https://github.com/kubernetes/enhancements/pull/6256

> - I don't think it actually helps the rollout problem. Focus for GA needs to be how we help people safely roll it out. If we decide the label is right, we can bring it back

I guess we should have a separate KEP to extend `nodeAffinity..matchFields` to allow looking up `NodeSystemInfo` properties

thockin commented:
/approve

dims commented:
/skip

AkihiroSuda commented:
/retest

dipesh-rawat commented:
Adding this back to the milestone based on the exception approval discussed in [Slack thread (here)](https://kubernetes.slack.com/archives/C2C40FMNF/p1784811064644969).


/milestone v1.37

dims commented:
/lgtm

AkihiroSuda commented:
Can anybody here also take a look at:
- https://github.com/kubernetes/website/pull/56420
- https://github.com/kubernetes/website/pull/56638
